src/awssh/autocompleter.py

```diff
@@ -0,0 +1,347 @@
+    log(f"command_line is {command_line}, command_index is {command_index}")
+    words: list[str]
+    cur_word: str
+    prev_word: str
+    words, cur_word, prev_word = process_command_line(command_line, command_index)
+    word_set: set[str] = set(words)
+
+    # Calculate the state from the current commandline
+    state = parse_command_line(words)
+    log(f"state is {state.__dict__}")
+
+    # Build a list of candidate completions
+    candidates: set[str] = set()
+
+    # If we are working on an option, then suggest parameters for that option.
+    if prev_word in CREDENTIALS_OPTIONS:
+        candidates |= get_cred_files()
+    elif prev_word in LOG_LEVEL_OPTIONS:
+        candidates |= LOG_LEVELS
+    elif prev_word in REGION_OPTIONS:
+        candidates |= get_regions()
+    else:
+        # Not working on an option parameter
+        if not state.profile:
+            # If we don't have the positional profile argument, suggest options
+            candidates |= build_option_candidates(word_set)
+
+        # If we have enough optional information start suggesting positional parameters
+        if state.cred_file.is_file() and state.aws_region:
+            profile_filter = os.environ.get("AWSSH_PROFILE_FILTER")
+            profiles = get_profiles(state.cred_file, profile_filter)
+            if state.profile not in profiles:
+                candidates |= profiles
+
+            # If we have enough information suggest instance ids and names
+            if not state.instance and state.profile in profiles:
+                instances: set[tuple[str, str]] = get_instances(
+                    state.cred_file, state.profile, state.aws_region
+                )
+                # Build sets of id-name pairs, and names alone
+                instance_id_and_name: set[str] = {f"{i[0]} {i[1]}" for i in instances}
+                instance_names: set[str] = {i[1] for i in instances}
+
+                # Include the names alone.  They will be substituted with an
+                # instance id when only one candidate is left.
+                candidates |= instance_names
+
+                # Filter instances now so we can strip off the name if there is only one
+                filtered_id_and_name = {
+                    i for i in instance_id_and_name if cur_word in i
+                }
+                if len(filtered_id_and_name) == 1:
+                    # If there is only one matching instance, remove the Name suffix
+                    candidates = {filtered_id_and_name.pop().split()[0]}
+                else:
+                    candidates |= instance_id_and_name
+
+    return candidates, cur_word
+
+
+def print_completions(comps: set[str], cur: str) -> None:
+    """Output the completions to stdout.
+
+    Args:
+        comps (set[str]): The completions to output.
+        cur (str): The current word.
+    """
+    log(f"completions: {sorted(comps)}")
+    contains: list[str] = list()
+    starts: list[str] = list()
```

Comment:
Generally using literals is preferred for things like this.

```suggestion
    contains: list[str] = []
    starts: list[str] = []
```